bbolt/bbolt.go

```diff
@@ -210,16 +195,36 @@ func (t bboltShelf) Get(key []byte) ([]byte, error) {
 	return append(value[:0:0], value...), nil
 }
 
-func (t bboltShelf) Put(key []byte, value []byte) error {
-	return t.bucket.Put(key, value)
+func (t bboltShelf) Put(key stoabs.Key, value []byte) error {
+	return t.bucket.Put(key.Bytes(), value)
 }
 
-func (t bboltShelf) Delete(key []byte) error {
-	return t.bucket.Delete(key)
+func (t bboltShelf) Delete(key stoabs.Key) error {
+	return t.bucket.Delete(key.Bytes())
 }
 
 func (t bboltShelf) Stats() stoabs.ShelfStats {
 	return stoabs.ShelfStats{
 		NumEntries: uint(t.bucket.Stats().KeyN),
 	}
 }
+
+func (t bboltShelf) Iterate(callback stoabs.CallerFn) error {
+	cursor := t.bucket.Cursor()
+	for k, v := cursor.First(); k != nil; k, v = cursor.Next() {
+		if err := callback(stoabs.BytesKey(k), v); err != nil {
```

Comment:
whatever you put into it, you always get the key back as bytes. Not sure how often you'd be using the key when reading (only when ranging/iterating), but I guess there's no ideal way to return the key in its type it was stored as?

Deriving from key length (4 bytes = int, 32 = 256bitkey, other = bytes) is ofc very naive. Adding a byte for every key indicating its type?

Or given we're not going to mix types, store it in BBolt itself? (but then we'd have to check it when you write data).

On second/third thought, maybe not worth the effort at this point.

Reply:
The challenge is that every key in Redis is converted to a string, when requesting the Next key in line this can be challenging if the key is supposed to be a string. Hence the helpers

Reply:
Everything from bbolt is bytes, different implementations use different types of keys. We'll have to see how it works out in practise.